### PR TITLE
Add permission granting/revoking

### PR DESCRIPTION
**Pull request by leventeliu**

### CODE REVIEW

File: blockproducer/metastate.go [hunk 1]
```diff
@@ -1017,29 +1035,68 @@ func (s *metaState) transferSQLChainTokenBalance(transfer *types.Transfer) (err
...
+				if user.Arrears <= transfer.Amount {
+					for _, miner := range sqlchain.Miners {
+						newUserArrears := make([]*types.UserArrears, len(miner.UserArrears))
+						i := 0
+						for _, ua := range miner.UserArrears {
+							if ua.User == user.Address {
+								miner.PendingIncome += miner.UserArrears[i].Arrears
+							} else {
+								newUserArrears[i] = ua
+								i += 1
```
Comment (covenantdb) [inline blockproducer/metastate.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
should replace i += 1 with i++

File: worker/chainbusservice.go [hunk 2]
```diff
@@ -179,12 +191,22 @@ func (bs *BusService) requestLastBlock() (
...
 func (bs *BusService) RequestSQLProfile(dbID proto.DatabaseID) (p *types.SQLChainProfile, ok bool) {
-	bs.lock.Lock()
-	defer bs.lock.Unlock()
+	bs.lock.RLock()
+	defer bs.lock.RUnlock()
 	p, ok = bs.sqlChainProfiles[dbID]
 	return
 }
 
+func (bs *BusService) RequestPermStat(dbID proto.DatabaseID, user proto.AccountAddress) (permStat *types.PermStat, ok bool) {
```
Comment (covenantdb) [inline worker/chainbusservice.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method BusService.RequestPermStat should have comment or be unexported

File: types/token.go [hunk 3]
```diff
@@ -45,6 +47,19 @@ var TokenList = map[TokenType]string{
 	Bitcoin:  "Bitcoin",
 }
 
+// MarshalJSON implements the json.Marshaler interface.
+func (t TokenType) MarshalJSON() ([]byte, error) {
+	res := make([]byte, 8)
+	binary.BigEndian.PutUint64(res, uint64(t))
+	return res, nil
```
Comment (xq262144) [inline types/token.go]: Better replaced with `return json.Marshal(t.String())`

File: types/token.go [hunk 4]
```diff
@@ -45,6 +47,19 @@ var TokenList = map[TokenType]string{
...
+}
+
+// UnmarshalJSON implements the json.Unmarshaler interface.
+func (t *TokenType) UnmarshalJSON(data []byte) (err error) {
+	*t = TokenType(binary.BigEndian.Uint64(data))
+	return
+}
```
Comment (xq262144) [inline types/token.go]: Better replaced with
```go
var typeString string
if err = json.Unmarshal(data, &typeString); err != nil {
  return
}
*t = FromString(typeString)

return
```

File: worker/dbms.go [hunk 5]
```diff
@@ -47,7 +47,7 @@ const (
 	DBMetaFileName = "db.meta"
 
 	// CheckInterval defines the bus service period.
-	CheckInterval = time.Second
+	CheckInterval = 30 * time.Second
```
Comment (xq262144) [inline worker/dbms.go]: Why change this?

File: blockproducer/metastate.go [hunk 6]
```diff
@@ -1105,15 +1150,22 @@ func (s *metaState) transferSQLChainTokenBalance(transfer *types.Transfer) (err
...
 				err = safeAdd(&user.AdvancePayment, &transfer.Amount)
 				if err != nil {
-					return err
+					return
+				}
+			}
+			account.TokenBalance[transfer.TokenType] -= transfer.Amount
+			if !user.Status.EnableQuery() {
+				if user.AdvancePayment > minDep {
+					user.Status = types.Normal
```
Comment (xq262144) [inline blockproducer/metastate.go]: Should we move `minDep` value from `AdvancePayment` to `Deposit` ?

File: blockproducer/metastate.go [hunk 7]
```diff
@@ -1074,29 +1066,82 @@ func (s *metaState) transferSQLChainTokenBalance(transfer *types.Transfer) (err
...
+			"sender": transfer.Sender.String(),
+		}).WithError(err).Warning("database not exist in transferSQLChainTokenBalance")
+		return
+	}
+	if sqlchain.TokenType != transfer.TokenType {
+		err = ErrWrongTokenType
+		log.WithFields(log.Fields{
+			"dbid":   transfer.Receiver.DatabaseID(),
+			"sender": transfer.Sender.String(),
+		})
```
Comment (auxten) [inline blockproducer/metastate.go]: Missing Error()

File: client/driver.go [hunk 8]
```diff
@@ -236,40 +237,136 @@ func GetStableCoinBalance() (balance uint64, err error) {
...
+	}
 
-	var pubKey *asymmetric.PublicKey
+	nonce, err = getNonce(addr)
+	if err != nil {
+		return
+	}
+
+	up := types.NewUpdatePermission(&types.UpdatePermissionHeader{
+		TargetSQLChain: targetChain,
```
Comment (xq262144) [inline client/driver.go:283]: May be `TargetSQLChain` should be type `proto.DatabaseID` ?

**Review by xq262144 — APPROVED** [review]
Approve

File: blockproducer/metastate.go [hunk 9]
```diff
@@ -903,9 +888,13 @@ func (s *metaState) updatePermission(tx *types.UpdatePermission) (err error) {
 
 	// check whether sender is admin and find targetUser
 	isAdmin := false
+	numOfAdmin := 0
 	targetUserIndex := -1
 	for i, u := range so.Users {
 		isAdmin = isAdmin || (sender == u.Address && u.Permission == types.Admin)
+		if u.Permission == types.Admin {
+			numOfAdmin += 1
```
Comment (covenantdb) [inline blockproducer/metastate.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
should replace numOfAdmin += 1 with numOfAdmin++

**Review by auxten — APPROVED** [review]
LGTM

**Pull request merged**